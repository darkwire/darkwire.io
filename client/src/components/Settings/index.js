import React, { Component } from 'react';
import PropTypes from 'prop-types';
import RoomLink from 'components/RoomLink';
import { styles } from './styles.module.scss';
import Cookie from 'js-cookie';
import T from 'components/T';

class Settings extends Component {
  handleSoundToggle() {
    this.props.toggleSoundEnabled(!this.props.soundIsEnabled);
  }

  handleNotificationToggle() {
    Notification.requestPermission().then(permission => {
      if (permission === 'granted') {
        this.props.toggleNotificationEnabled(!this.props.notificationIsEnabled);
        this.props.toggleNotificationAllowed(true);
      }
      if (permission === 'denied') {
        this.props.toggleNotificationAllowed(false);
      }
    });
  }

  handleLanguageChange(evt) {
    const language = evt.target.value;
    Cookie.set('language', language);
    this.props.setLanguage(language);
  }

  render() {
    return (
      <div className={styles}>
        <section>
          <h4>
            <T path="newMessageNotification" />
          </h4>
          <form>
            <div className="form-check">
              <label className="form-check-label" htmlFor="sound-control">
                <input
                  id="sound-control"
                  onChange={this.handleSoundToggle.bind(this)}
                  className="form-check-input"
                  type="checkbox"
                  checked={this.props.soundIsEnabled}
                />
                <T path="sound" />
              </label>
            </div>
            <div className="form-check">
              <label className="form-check-label" htmlFor="notif-control">
                {this.props.notificationIsAllowed !== false && 
                  <>
                    <input
                      id="notif-control"
                      onChange={this.handleNotificationToggle.bind(this)}
                      className="form-check-input"
                      type="checkbox"
                      checked={this.props.notificationIsEnabled}
                      disabled={this.props.notificationIsAllowed === false} // Important to keep '=== false' here
                    />
                    <T path="desktopNotification" />
                  </>
                }
                {this.props.notificationIsAllowed === false && <T path="desktopNotificationBlocked" />}
              </label>
            </div>
          </form>
        </section>

        <section>
          <h4 className="mb-3">
            <T path="copyRoomHeader" />
          </h4>
          <RoomLink roomId={this.props.roomId} translations={this.props.translations} />
        </section>

        <section>
          <h4 className="mb-3">
            <T path="languageDropdownHeader" />
          </h4>
          <p>
            <a
              href="https://github.com/darkwire/darkwire.io/blob/master/client/README.md#translations"
              rel="noopener noreferrer"
              target="_blank"
            >
              <T path="helpTranslate" />
            </a>
          </p>
          <div className="form-group">
            <select
              value={this.props.language}
              className="form-control"
              onChange={this.handleLanguageChange.bind(this)}
            >
              <option value="en">English</option>
              <option value="fr">Français</option>
              <option value="oc">Occitan</option>
              <option value="de">Deutsch</option>
              <option value="nl">Nederlands</option>
              <option value="it">Italiano</option>
              <option value="zhCN">中文</option>
            </select>
          </div>
        </section>

        <section>
          <h4>
            <T path="roomOwnerHeader" />
          </h4>
          <p>
            <T path="roomOwnerText" />
          </p>
        </section>
        <section>
          <h4>
            <T path="lockRoomHeader" />
          </h4>
          <p>
            <T path="lockRoomText" />
          </p>
        </section>
        <section>
          <h4>
            <T path="slashCommandsHeader" />
          </h4>
          <p>
            <T path="slashCommandsText" />
          </p>
          <ul>
            <li>
              /nick [username]{' '}
              <span className="text-muted">
                <T path="slashCommandsBullets.0" />
              </span>
            </li>
            <li>
              /me [action]{' '}
              <span className="text-muted">
                <T path="slashCommandsBullets.1" />
              </span>
            </li>
            <li>
              /clear{' '}
              <span className="text-muted">
                <T path="slashCommandsBullets.2" />
              </span>
            </li>
            <li>
              /help{' '}
              <span className="text-muted">
                <T path="slashCommandsBullets.3" />
              </span>
            </li>
          </ul>
        </section>
      </div>
    );
  }
}

Settings.propTypes = {
  soundIsEnabled: PropTypes.bool.isRequired,
  toggleSoundEnabled: PropTypes.func.isRequired,
  notificationIsEnabled: PropTypes.bool.isRequired,
  notificationIsAllowed: PropTypes.bool,
  toggleNotificationEnabled: PropTypes.func.isRequired,
  toggleNotificationAllowed: PropTypes.func.isRequired,
  roomId: PropTypes.string.isRequired,
  setLanguage: PropTypes.func.isRequired,
  translations: PropTypes.object.isRequired,
};

export default Settings;
