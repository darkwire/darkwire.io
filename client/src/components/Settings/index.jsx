import PropTypes from 'prop-types';

import RoomLink from '@/components/RoomLink';
import T from '@/components/T';

import classes from './styles.module.scss';

const Settings = ({
  soundIsEnabled,
  persistenceIsEnabled,
  toggleSoundEnabled,
  notificationIsEnabled,
  notificationIsAllowed,
  toggleNotificationEnabled,
  toggleNotificationAllowed,
  togglePersistenceEnabled,
  roomId,
  setLanguage,
  language,
  translations,
}) => {
  const handleSoundToggle = () => {
    toggleSoundEnabled(!soundIsEnabled);
  };

  const handlePersistenceToggle = () => {
    togglePersistenceEnabled(!persistenceIsEnabled);
  };

  const handleNotificationToggle = () => {
    Notification.requestPermission().then(permission => {
      if (permission === 'granted') {
        toggleNotificationEnabled(!notificationIsEnabled);
        toggleNotificationAllowed(true);
      }
      if (permission === 'denied') {
        toggleNotificationAllowed(false);
      }
    });
  };

  const handleLanguageChange = evt => {
    setLanguage(evt.target.value);
  };

  return (
    <div className={classes.styles}>
      <section>
        <h4>
          <T path="newMessageNotification" />
        </h4>
        <form>
          <div className="form-check">
            <label className="form-check-label" htmlFor="sound-control">
              <input
                id="sound-control"
                onChange={handleSoundToggle}
                className="form-check-input"
                type="checkbox"
                checked={soundIsEnabled}
              />
              <T path="sound" />
            </label>
          </div>
          <div className="form-check">
            <label className="form-check-label" htmlFor="notif-control">
              {notificationIsAllowed !== false && (
                <>
                  <input
                    id="notif-control"
                    onChange={handleNotificationToggle}
                    className="form-check-input"
                    type="checkbox"
                    checked={notificationIsEnabled}
                    disabled={notificationIsAllowed === false} // Important to keep '=== false' here
                  />
                  <T path="desktopNotification" />
                </>
              )}
              {notificationIsAllowed === false && <T path="desktopNotificationBlocked" />}
            </label>
          </div>
          <div className="form-check">
            <label className="form-check-label" htmlFor="persistence-control">
              <input
                id="persistence-control"
                onChange={handlePersistenceToggle}
                className="form-check-input"
                type="checkbox"
                checked={persistenceIsEnabled}
              />
              <T path="persistence" />
            </label>
          </div>
        </form>
      </section>

      <section>
        <h4 className="mb-3">
          <T path="copyRoomHeader" />
        </h4>
        <RoomLink roomId={roomId} translations={translations} />
      </section>

      <section>
        <h4 className="mb-3">
          <T path="languageDropdownHeader" />
        </h4>
        <p>
          <a
            href="https://github.com/darkwire/darkwire.io/blob/master/client/README.md#translations"
            rel="noopener noreferrer"
            target="_blank"
          >
            <T path="helpTranslate" />
          </a>
        </p>
        <div className="form-group">
          <select value={language} className="form-control" onChange={handleLanguageChange}>
            <option value="en">English</option>
            <option value="fr">Français</option>
            <option value="oc">Occitan</option>
            <option value="de">Deutsch</option>
            <option value="esAR">Español (Argentina)</option>
            <option value="nl">Nederlands</option>
            <option value="it">Italiano</option>
            <option value="ru">Русский</option>
            <option value="pl">Polish</option>
            <option value="zhCN">中文</option>
            <option value="ja">日本語</option>
            <option value="tr">Türkçe</option>
            <option value="ko">한국어</option>
          </select>
        </div>
      </section>

      <section>
        <h4>
          <T path="roomOwnerHeader" />
        </h4>
        <p>
          <T path="roomOwnerText" />
        </p>
      </section>
      <section>
        <h4>
          <T path="lockRoomHeader" />
        </h4>
        <p>
          <T path="lockRoomText" />
        </p>
      </section>
      <section>
        <h4>
          <T path="slashCommandsHeader" />
        </h4>
        <p>
          <T path="slashCommandsText" />
        </p>
        <ul>
          <li>
            /nick [username]{' '}
            <span className="text-muted">
              <T path="slashCommandsBullets.0" />
            </span>
          </li>
          <li>
            /me [action]{' '}
            <span className="text-muted">
              <T path="slashCommandsBullets.1" />
            </span>
          </li>
          <li>
            /clear{' '}
            <span className="text-muted">
              <T path="slashCommandsBullets.2" />
            </span>
          </li>
          <li>
            /help{' '}
            <span className="text-muted">
              <T path="slashCommandsBullets.3" />
            </span>
          </li>
        </ul>
      </section>
    </div>
  );
};

Settings.propTypes = {
  soundIsEnabled: PropTypes.bool.isRequired,
  persistenceIsEnabled: PropTypes.bool.isRequired,
  toggleSoundEnabled: PropTypes.func.isRequired,
  togglePersistenceEnabled: PropTypes.func.isRequired,
  notificationIsEnabled: PropTypes.bool.isRequired,
  notificationIsAllowed: PropTypes.bool,
  toggleNotificationEnabled: PropTypes.func.isRequired,
  toggleNotificationAllowed: PropTypes.func.isRequired,
  roomId: PropTypes.string.isRequired,
  setLanguage: PropTypes.func.isRequired,
  language: PropTypes.string.isRequired,
  translations: PropTypes.object.isRequired,
};

export default Settings;
