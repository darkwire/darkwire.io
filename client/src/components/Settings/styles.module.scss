.styles {
  section {
    margin-bottom: 40px;
  }
}
