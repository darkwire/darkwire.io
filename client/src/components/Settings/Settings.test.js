import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import Settings from '.';

const mockTranslations = {
  sound: 'soundCheck',
};

jest.mock('components/T', () => {
  return jest.fn().mockImplementation(({ path }) => {
    return mockTranslations[path] || 'default';
  });
});

// To avoid missing provider
jest.mock('components/T');
jest.mock('components/RoomLink');

describe('Settings component', () => {
  it('should display', async () => {
    const { asFragment } = render(
      <Settings
        soundIsEnabled={true}
        toggleSoundEnabled={() => {}}
        roomId="roomId"
        setLanguage={() => {}}
        translations={mockTranslations}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should toggle sound', async () => {
    const toggleSound = jest.fn();
    const { getAllByText } = render(
      <Settings
        soundIsEnabled={true}
        toggleSoundEnabled={toggleSound}
        roomId="roomId"
        setLanguage={() => {}}
        translations={mockTranslations}
      />,
    );

    //console.log(getAllByText(mockTranslations.sound)[1]);
    fireEvent.click(getAllByText(mockTranslations.sound)[1]);

    expect(toggleSound).toHaveBeenCalledWith(false);
  });

  it('should change lang', async () => {
    const changeLang = jest.fn();

    const { getByDisplayValue } = render(
      <Settings
        soundIsEnabled={true}
        toggleSoundEnabled={() => {}}
        roomId="roomId"
        setLanguage={changeLang}
        translations={{}}
      />,
    );

    fireEvent.change(getByDisplayValue('English'), { target: { value: 'de' } });

    expect(changeLang).toHaveBeenCalledWith('de');
  });
});
