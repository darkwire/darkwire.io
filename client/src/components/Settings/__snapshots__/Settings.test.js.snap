// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings component should display 1`] = `
<DocumentFragment>
  <div
    class="styles"
  >
    <section>
      <h4>
        New message notification
      </h4>
      <form>
        <div
          class="form-check"
        >
          <label
            class="form-check-label"
            for="sound-control"
          >
            <input
              checked=""
              class="form-check-input"
              id="sound-control"
              type="checkbox"
            />
            Sound
          </label>
        </div>
        <div
          class="form-check"
        >
          <label
            class="form-check-label"
            for="notif-control"
          >
            <input
              checked=""
              class="form-check-input"
              id="notif-control"
              type="checkbox"
            />
            Desktop Notification
          </label>
        </div>
      </form>
    </section>
    <section>
      <h4
        class="mb-3"
      >
        This room
      </h4>
    </section>
    <section>
      <h4
        class="mb-3"
      >
        Language
      </h4>
      <p>
        <a
          href="https://github.com/darkwire/darkwire.io/blob/master/client/README.md#translations"
          rel="noopener noreferrer"
          target="_blank"
        >
          Help us translate Darkwire!
        </a>
      </p>
      <div
        class="form-group"
      >
        <select
          class="form-control"
        >
          <option
            value="en"
          >
            English
          </option>
          <option
            value="fr"
          >
            Français
          </option>
          <option
            value="oc"
          >
            Occitan
          </option>
          <option
            value="de"
          >
            Deutsch
          </option>
          <option
            value="nl"
          >
            Nederlands
          </option>
          <option
            value="it"
          >
            Italiano
          </option>
          <option
            value="zhCN"
          >
            中文
          </option>
        </select>
      </div>
    </section>
    <section>
      <h4>
        Room Ownership
      </h4>
      <p>
        The person who created the room is the room owner and has special privileges, like the ability to lock and unlock the room. If the owner leaves the room, the second person to join assumes ownership. If they leave, the third person becomes owner, and so on. The room owner has a star icon next to their username in the participants dropdown.
      </p>
    </section>
    <section>
      <h4>
        Lock Room
      </h4>
      <p>
        If you are the room owner, you can lock and unlock the room by clicking the lock icon in the nav bar. When a room is locked, no other participants will be able to join.
      </p>
    </section>
    <section>
      <h4>
        Slash Commands
      </h4>
      <p>
        The following slash commands are available:
      </p>
      <ul>
        <li>
          /nick [username] 
          <span
            class="text-muted"
          >
            changes username
          </span>
        </li>
        <li>
          /me [action] 
          <span
            class="text-muted"
          >
            performs an action
          </span>
        </li>
        <li>
          /clear 
          <span
            class="text-muted"
          >
            clears your message history
          </span>
        </li>
        <li>
          /help 
          <span
            class="text-muted"
          >
            lists all commands
          </span>
        </li>
      </ul>
    </section>
  </div>
</DocumentFragment>
`;

exports[`Settings component should display 2`] = `
<DocumentFragment>
  <div
    class="styles"
  >
    <section>
      <h4>
        New message notification
      </h4>
      <form>
        <div
          class="form-check"
        >
          <label
            class="form-check-label"
            for="sound-control"
          >
            <input
              checked=""
              class="form-check-input"
              id="sound-control"
              type="checkbox"
            />
            Sound
          </label>
        </div>
        <div
          class="form-check"
        >
          <label
            class="form-check-label"
            for="notif-control"
          >
            Desktop notifications have been denied
          </label>
        </div>
      </form>
    </section>
    <section>
      <h4
        class="mb-3"
      >
        This room
      </h4>
    </section>
    <section>
      <h4
        class="mb-3"
      >
        Language
      </h4>
      <p>
        <a
          href="https://github.com/darkwire/darkwire.io/blob/master/client/README.md#translations"
          rel="noopener noreferrer"
          target="_blank"
        >
          Help us translate Darkwire!
        </a>
      </p>
      <div
        class="form-group"
      >
        <select
          class="form-control"
        >
          <option
            value="en"
          >
            English
          </option>
          <option
            value="fr"
          >
            Français
          </option>
          <option
            value="oc"
          >
            Occitan
          </option>
          <option
            value="de"
          >
            Deutsch
          </option>
          <option
            value="nl"
          >
            Nederlands
          </option>
          <option
            value="it"
          >
            Italiano
          </option>
          <option
            value="zhCN"
          >
            中文
          </option>
        </select>
      </div>
    </section>
    <section>
      <h4>
        Room Ownership
      </h4>
      <p>
        The person who created the room is the room owner and has special privileges, like the ability to lock and unlock the room. If the owner leaves the room, the second person to join assumes ownership. If they leave, the third person becomes owner, and so on. The room owner has a star icon next to their username in the participants dropdown.
      </p>
    </section>
    <section>
      <h4>
        Lock Room
      </h4>
      <p>
        If you are the room owner, you can lock and unlock the room by clicking the lock icon in the nav bar. When a room is locked, no other participants will be able to join.
      </p>
    </section>
    <section>
      <h4>
        Slash Commands
      </h4>
      <p>
        The following slash commands are available:
      </p>
      <ul>
        <li>
          /nick [username] 
          <span
            class="text-muted"
          >
            changes username
          </span>
        </li>
        <li>
          /me [action] 
          <span
            class="text-muted"
          >
            performs an action
          </span>
        </li>
        <li>
          /clear 
          <span
            class="text-muted"
          >
            clears your message history
          </span>
        </li>
        <li>
          /help 
          <span
            class="text-muted"
          >
            lists all commands
          </span>
        </li>
      </ul>
    </section>
  </div>
</DocumentFragment>
`;
